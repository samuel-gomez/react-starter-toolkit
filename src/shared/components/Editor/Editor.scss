@import '@axa-fr/react-toolkit-core/dist/assets/scss/core.scss';

@mixin editPropLink {
  margin-left: auto;
  text-decoration: none;
  cursor: pointer;
  * {
    color: $color-red-error;
  }
}

.af-editor {
  display: flex;
  flex-direction: column;
  margin-bottom: 2rem;
  &__header {
    background-color: black;
    padding: 0.5rem 1rem;
    color: $color-gray-2;
    display: flex;
    gap: 1rem;
    line-height: 1;
    .af-link {
      color: $color-gray-2;
      display: flex;
      height: 18px;
      fill: $color-gray-2;
      line-height: 1;
      &:has(.glyphicon.glyphicon-cog) {
        @include editPropLink;
      }
      .af-icon-svg {
        margin-right: 0.5rem;
        height: 100%;
      }
      .glyphicon-cog {
        margin-right: 0.4rem;
        text-decoration: none;
      }
      &__npm-name {
        position: relative;
        width: 126px;
        height: 20px;
      }
    }
  }
}

.af-btn--circle--close-editor {
  background: $color-axa;
  color: white;
  width: 26px;
  height: 26px;
  font-size: 1rem;
  border-color: $color-axa;
  transition: background-color 0.3s linear, color 0.3s linear;
  &:hover {
    background: white;
    color: $color-axa;
  }
}

@supports not selector(:has) {
  .af-editor {
    &__header {
      .af-link:last-child {
        @include editPropLink;
      }
    }
  }
}

@include media-breakpoint-down(sm) {
  .af-editor {
    margin-bottom: 0;
    &__header {
      flex-wrap: wrap;
      justify-content: space-between;
    }
  }
}

.af-form-editor {
  display: grid;
  grid-template-columns: 1fr;
  background-color: $color-white;
  padding-bottom: 1.5rem;
  &__field {
    padding: 0.4rem 1rem;
    &:has(.af-form-editor__separator) {
      padding: 0.2rem;
      background: $color-gray-2;
      color: $color-scorpion;
      margin: 1rem 0 0 0;
    }
    &:not(:last-child) {
      border-bottom: 1px solid $color-gray-2;
    }
    & > label {
      display: block;
      width: 100%;
    }
    .af-form__group-label {
      color: $color-scorpion;
      font-size: 0.8rem;
    }
    .af-form__input-text {
      padding: 0.2rem 0.5rem;
      width: 100%;
    }
    .af-form__select-container {
      display: block;
    }
    .af-form__input-select {
      padding: 0.2rem 2.7rem 0.2rem 1rem;
      width: 100%;
    }
    .af-form__input-textarea {
      width: 100%;
    }
    .af-form-editor__separator {
      margin: 0;
    }
  }
  .af-form__group {
    &--date-custom {
      margin-bottom: 0;
    }
  }
}
.af-btn--editor {
  padding: 0.5rem 1rem;
  min-width: auto;
  border-radius: 0.5rem;
}

.af-draggable {
  display: flex;
  flex-direction: column;
  width: 300px;
  box-shadow: 2px 2px 10px 0 rgba(0, 0, 0, 0.2);
  position: relative;
  &:after {
    content: '';
    position: absolute;
    width: 100%;
    height: 13px;
    bottom: 0;
    left: 0;
    box-shadow: inset 0px -11px 12px -7px rgb(0 0 0 / 56%);
    z-index: 2;
  }
  &-container {
    position: fixed;
    top: 20px;
    right: 20px;
    z-index: 2001;
    display: flex;
  }

  &__title {
    background-color: $color-axa;
    color: white;
    padding: 0.4rem 1rem;
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 0;
    cursor: grab;
  }
  &__tools {
    display: flex;
    gap: 1rem;
    align-items: center;
  }
  &__content {
    overflow-y: scroll;
    overflow-x: hidden;
    resize: vertical;
    max-height: 90vh;
    background-color: white;
  }
  .glyphicon-move {
    font-size: 1.6rem;
  }
  .glyphicon-close {
    cursor: pointer;
  }
}
