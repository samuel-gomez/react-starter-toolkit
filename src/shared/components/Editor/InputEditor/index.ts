export { default } from './InputEditor';
export type { TInputEditor } from './InputEditor';
