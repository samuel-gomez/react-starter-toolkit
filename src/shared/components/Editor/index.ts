export { withEditor, useEditable } from './Editor';
export type { TEvent } from './Editor';
