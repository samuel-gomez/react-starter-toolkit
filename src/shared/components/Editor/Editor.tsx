import { ComponentType, FocusEvent, useCallback, useState } from 'react';
import Draggable from 'react-draggable';
import Icons from 'shared/components/Icons';
import { ClickEvent } from '@axa-fr/react-toolkit-core';
import { DESIGN_SYSTEM, GITHUB, STORYBOOK } from 'shared/constants';
import InputEditor, { TInputEditor } from './InputEditor';
import './Editor.scss';

export type TEvent = {
  value?: string;
  id?: string;
  name?: string;
  values?: string[];
};

const omittedProps = ['onChange', 'knobs', 'onClick', 'values'];

export type TonChange = (e: TEvent) => void;

export type Tknobs = Record<string, Record<string, TInputEditor['value']>>;

type TmergePropsAndKnobs<P> = {
  props: P & Record<string, unknown>;
  knobs: Tknobs;
};

export const mergePropsAndKnobs = <P extends object>({ props, knobs }: TmergePropsAndKnobs<P>) =>
  Object.keys(props).reduce(
    (acc, key) => ({
      ...acc,
      [key]: knobs[key] ? { ...knobs[key], value: props[key] } : props[key],
    }),
    {},
  );

type TLabelEditor = {
  value: TInputEditor['value'];
  name: string;
};
export const LabelEditor = ({ name, value }: TLabelEditor) =>
  !(typeof value === 'object' && !!value.type && value.type === 'separator') ? (
    <label className="af-form__group-label" htmlFor={name}>
      {name}
    </label>
  ) : (
    <></>
  );

type TFieldEditor = {
  onChange: TonChange;
  name: string;
  value: TInputEditor['value'];
};

const FieldEditor = ({ name, value, onChange }: TFieldEditor) => (
  <div className={`af-form-editor__field`} key={name}>
    <LabelEditor value={value} name={name} />
    <InputEditor key={name} name={name} value={value} onChange={onChange} />
  </div>
);

/**
 * @param onChange : function for update state fields value
 * @param knobs : overrided props (for object type)
 * @returns Form fields editor
 */
export const FormEditor = <P extends object>({
  onChange,
  knobs,
  ...props
}: Omit<P, 'onClick'> & { onChange: ReturnType<typeof useEditable>['onChange']; knobs: Tknobs }) => (
  <form className="af-form-editor">
    <>
      {Object.entries(mergePropsAndKnobs({ props, knobs }))
        .filter(([key, val]) => !omittedProps.includes(key) && typeof val !== 'function')
        .map(([name, value]) => (
          <FieldEditor key={name} name={name} value={value as TInputEditor['value']} onChange={onChange(name)} />
        ))}
    </>
  </form>
);

export const useToggleEditor = (initState = false) => {
  const [isOpenEditor, setStateToggleEditor] = useState(initState);

  const closeEditor = useCallback(() => {
    setStateToggleEditor(false);
  }, []);

  const openEditor = useCallback(() => {
    setStateToggleEditor(true);
  }, []);

  return { closeEditor, openEditor, isOpenEditor };
};

export type TReturnUseToggleEditor = ReturnType<typeof useToggleEditor>;

/**
 * HOC to add Editor
 * @param Component : component to be wrapped
 * @param knobs : for lists props values
 * @returns Component Editable
 */
export const withEditor =
  <P extends object>(Component: ComponentType<P>, knobs: Tknobs = {}, useToggleEditorFn = useToggleEditor): ComponentType<P> =>
  (props: P) => {
    const { closeEditor, openEditor, isOpenEditor } = useToggleEditorFn();
    return (
      <section className={'af-editor'}>
        <Component {...props} openEditor={openEditor} isOpenEditor={isOpenEditor} />
        {isOpenEditor && (
          <Draggable cancel=".glyphicon-close" handle=".af-draggable__title">
            <div className="af-draggable-container">
              <div className="af-draggable">
                <h3 className="af-draggable__title">
                  Props Editor
                  <div className="af-draggable__tools">
                    <i className="glyphicon glyphicon-move"></i>
                    <i className="glyphicon glyphicon-close" onClick={closeEditor}></i>
                  </div>
                </h3>

                <div className="af-draggable__content">
                  <FormEditor<P> {...(props as P & { onChange: ReturnType<typeof useEditable>['onChange'] })} knobs={knobs} />
                </div>
              </div>
            </div>
          </Draggable>
        )}
      </section>
    );
  };

export const setValue = (value: string) => {
  if (value === 'false') return true;
  if (value === 'true') return false;

  return value;
};

type TuseEditable<T> = {
  initialState: T;
  logEventFn?: typeof console.log;
  setValueFn?: typeof setValue;
};

export const useEditable = <T extends object>({ initialState, logEventFn = console.log, setValueFn = setValue }: TuseEditable<T>) => {
  const [state, setState] = useState(initialState);

  const onClick = useCallback(
    (name: string) => (e: ClickEvent) => {
      logEventFn('click event', name, e);
    },
    [logEventFn],
  );

  const onChange = useCallback(
    (key: string) => (e: TEvent) => {
      if (!e.value && e.values) {
        setState(prevState => ({ ...prevState, [key]: e.values }));
      } else {
        setState(prevState => ({ ...prevState, [key]: setValueFn(e?.value ?? '') }));
      }
    },
    [setValueFn],
  );

  const onBlur = useCallback(() => {
    setState(prevState => ({ ...prevState, autoFocus: false }));
  }, []);

  const onFocus = useCallback((e: FocusEvent<HTMLInputElement>) => {
    e.target.setSelectionRange(e.target.value.length, e.target.value.length);
    setState(prevState => ({ ...prevState, autoFocus: true }));
  }, []);

  return { onClick, onChange, state, onFocus, onBlur };
};

type TEditorHeader = {
  storybookPath?: string;
  designSystemPath?: string;
  githubPackage?: string;
  npmName?: string;
} & Partial<TReturnUseToggleEditor>;

export const EditorHeader = ({ storybookPath = '', designSystemPath = '', githubPackage = '', npmName = '', openEditor }: TEditorHeader) => (
  <header className="af-editor__header">
    {!!designSystemPath && (
      <a className="af-link" href={`${DESIGN_SYSTEM}${designSystemPath}`} target="_blank" rel="noopener noreferrer">
        <Icons icon="slash" viewBox="0 0 342.988 280" />
        Guidelines
      </a>
    )}

    {!!storybookPath && (
      <a className="af-link" href={`${STORYBOOK}${storybookPath}`} target="_blank" rel="noopener noreferrer">
        <Icons icon="storybook" viewBox="0 0 9.6 12" />
        Storybook
      </a>
    )}

    {!!githubPackage && (
      <a className="af-link" href={`${GITHUB}${githubPackage}`} target="_blank" rel="noopener noreferrer">
        <Icons icon="github" viewBox="0 0 438 438" />
        Github
      </a>
    )}

    {!!npmName && (
      <a target="_blank" rel="noopener noreferrer" href={`https://badge.fury.io/js/${npmName}`}>
        <img src={`https://badge.fury.io/js/${npmName}.svg`} alt="npm version" />
      </a>
    )}

    <span className="af-link" onClick={openEditor}>
      <i className="glyphicon glyphicon-cog"></i>
      <span className="af-link">Edit props</span>
    </span>
  </header>
);
