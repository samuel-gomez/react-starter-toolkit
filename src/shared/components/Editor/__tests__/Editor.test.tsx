import { FocusEvent } from 'react';
import { render } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import { Text } from '@axa-fr/react-toolkit-all';
import { mergePropsAndKnobs, withEditor, setValue, useEditable, useToggleEditor, TEvent } from '../Editor';

describe('setValue', () => {
  it.each`
    value      | expected
    ${'true'}  | ${false}
    ${'false'} | ${true}
    ${'value'} | ${'value'}
  `('Should return expected: $expected when value: $value', ({ value, expected }) => {
    const result = setValue(value);
    expect(result).toEqual(expected);
  });
});

describe('mergePropsAndKnobs', () => {
  const otherProps = {
    className: 'af-btn',
    label: 'valider',
    disabled: false,
    icon: '',
  };
  const props = {
    ...otherProps,
    classModifier: 'error',
  };
  const knobs = {
    classModifier: {
      value: '',
      options: [
        {
          label: 'Success',
          value: 'success',
        },
      ],
    },
  };
  const expectedMerge = { ...otherProps, classModifier: { ...knobs.classModifier, value: 'error' } };

  it.each`
    props    | knobs    | expected
    ${{}}    | ${{}}    | ${{}}
    ${props} | ${knobs} | ${expectedMerge}
  `('Should return expected: $expected when props: $props, knobs: $knobs', ({ props, knobs, expected }) => {
    const result = mergePropsAndKnobs({ props, knobs });
    expect(result).toEqual(expected);
  });
});

describe('useEditable', () => {
  const initialState = {
    classModifier: '',
    className: 'af-btn',
    label: 'valider',
    disabled: false,
    icon: '',
    autoFocus: false,
    values: [],
  };

  it('Should logEventFn have been called when onClick have been called', () => {
    const logEventFn = jest.fn();
    const { result } = renderHook(() => useEditable({ initialState, logEventFn }));
    act(() => result.current.onClick('name')({ id: 'id' }));
    expect(logEventFn).toHaveBeenCalledWith('click event', 'name', { id: 'id' });
  });

  it('Should setValueFn have been called and return updated state value when onChange have been called with event with value', () => {
    const setValueFn = jest.fn().mockReturnValue('setted value');
    const { result } = renderHook(() => useEditable({ initialState, setValueFn }));
    act(() => result.current.onChange('classModifier')({ value: 'newvalue' }));
    expect(setValueFn).toHaveBeenCalledWith('newvalue');
    expect(result.current.state.classModifier).toEqual('setted value');
  });

  it('Should setValueFn have been called with empty string and return updated state value when onChange have been called with event with undefined value', () => {
    const setValueFn = jest.fn().mockReturnValue('setted value');
    const { result } = renderHook(() => useEditable({ initialState, setValueFn }));
    act(() => result.current.onChange('classModifier')({ value: undefined }));
    expect(setValueFn).toHaveBeenCalledWith('');
    expect(result.current.state.classModifier).toEqual('setted value');
  });

  it('Should return updated state values when onChange have been called with event with values', () => {
    const { result } = renderHook(() => useEditable({ initialState }));
    act(() => result.current.onChange('values')({ values: ['newvalue'] }));

    expect(result.current.state.values).toEqual(['newvalue']);
  });

  it('Should set autofocus state to false when onBlur have been called', () => {
    const { result } = renderHook(() => useEditable({ initialState: { autoFocus: true } }));
    act(() => result.current.onBlur());
    expect(result.current.state.autoFocus).toBeFalsy();
  });

  it('Should set autofocus state to true when onBlur have been called', () => {
    const { result } = renderHook(() => useEditable({ initialState: { autoFocus: false } }));
    const event = {
      preventDefault: jest.fn(),
      target: { value: '<p>the-new-value</p>', setSelectionRange: jest.fn() },
    } as unknown as FocusEvent<HTMLInputElement>;

    act(() => result.current.onFocus(event));
    expect(result.current.state.autoFocus).toBeTruthy();
  });
});

describe('withEditor', () => {
  type Props = { className?: string; onChange: ReturnType<typeof useEditable>['onChange'] };
  const onChange = jest.fn().mockImplementation(name => (e: TEvent) => {
    console.log('onChange');
  });
  const Component = ({ className, onChange }: Props) => (
    <Text id="idtext" name="nametext" type="text" value="hello" onChange={onChange} className={className} />
  );

  it('Should render Component with FormEditor when apply withEditor HOC and isOpenEditor true', () => {
    const useToggleEditorFn = jest.fn().mockReturnValue({
      closeEditor: jest.fn(),
      openEditor: jest.fn(),
      isOpenEditor: true,
    });
    const ComponentWithEditor = withEditor<Props>(Component, {}, useToggleEditorFn);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render Component with FormEditor when apply withEditor HOC', () => {
    const ComponentWithEditor = withEditor<Props>(Component);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  const useToggleEditorFnNoOpen = jest.fn().mockReturnValue({
    closeEditor: jest.fn(),
    openEditor: jest.fn(),
    isOpenEditor: false,
  });
  it("Shouldn't render Component with isOpenEditor false", () => {
    const ComponentWithEditor = withEditor<Props>(Component, {}, useToggleEditorFnNoOpen);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useToggleEditor', () => {
  it('Should isOpenEditor to be false when closeEditor called', () => {
    const { result } = renderHook(() => useToggleEditor());
    act(() => result.current.closeEditor());
    const resIsOpenEditor = result.current.isOpenEditor;
    expect(resIsOpenEditor).toBe(false);
  });

  it('Should isOpenEditor to be true when openEditor called', () => {
    const { result } = renderHook(() => useToggleEditor());
    act(() => result.current.openEditor());
    const resIsOpenEditor = result.current.isOpenEditor;
    expect(resIsOpenEditor).toBe(true);
  });
});
