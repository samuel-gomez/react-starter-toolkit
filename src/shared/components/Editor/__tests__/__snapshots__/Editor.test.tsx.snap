// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withEditor Should render Component with FormEditor when apply withEditor HOC 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
  </section>
</DocumentFragment>
`;

exports[`withEditor Should render Component with FormEditor when apply withEditor HOC and isOpenEditor true 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
    <div
      class="af-draggable-container react-draggable"
      style="transform: translate(0px,0px);"
    >
      <div
        class="af-draggable"
      >
        <h3
          class="af-draggable__title"
        >
          Props Editor
          <div
            class="af-draggable__tools"
          >
            <i
              class="glyphicon glyphicon-move"
            />
            <i
              class="glyphicon glyphicon-close"
            />
          </div>
        </h3>
        <div
          class="af-draggable__content"
        >
          <form
            class="af-form-editor"
          >
            <div
              class="af-form-editor__field"
            >
              <label
                class="af-form__group-label"
                for="className"
              >
                className
              </label>
              <input
                class="af-form__input-text"
                id="className"
                name="className"
                type="text"
                value="af-component"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`withEditor Shouldn't render Component with isOpenEditor false 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
  </section>
</DocumentFragment>
`;
